anndata/tests/helpers.py

```diff
@@ -576,3 +577,17 @@ def as_dense_dask_array(a):
 @as_dense_dask_array.register(sparse.spmatrix)
 def _(a):
     return as_dense_dask_array(a.toarray())
+
+
+def check_error_or_notes_match(e: pytest.ExceptionInfo, pattern: str):
+    """
+    Checks whether the printed error message or the notes contains the given pattern.
+
+    DOES NOT WORK IN IPYTHON - because of the way IPython handles exceptions
```

Comment:
It's just based on observation, so no link to add. This isn't exactly unexpected, since `ipython` overrides exception handling.

Reply:
So it monkeypatches `traceback.format_exception_only`?

Reply:
I don't know the specifics

Reply:
The PR you mentioned in https://github.com/ipython/ipython/issues/13849#issuecomment-1600677231 is now merged and IPython supports `__notes__`. Maybe that could be added here.

Reply:
I don't think that has an explanation for this. That specific issue is fixed. This still doesn't work in IPython.

```
Python 3.9.16 (main, Dec  7 2022, 10:15:13) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from __future__ import annotations
   ...: 
   ...: import re
   ...: import sys
   ...: import traceback
   ...: import pytest
   ...: 
   ...: def check_error_or_notes_match(e: pytest.ExceptionInfo, pattern: str | re.Pattern):
   ...:     """
   ...:     Checks whether the printed error message or the notes contains the given pattern.
   ...: 
   ...:     DOES NOT WORK IN IPYTHON - because of the way IPython handles exceptions
   ...:     """
   ...:     import traceback
   ...: 
   ...:     message = "".join(traceback.format_exception_only(e.type, e.value))
   ...:     assert re.search(
   ...:         pattern, message
   ...:     ), f"Could not find pattern: '{pattern}' in error:\n\n{message}\n"
   ...: 
   ...: 
   ...: def add_note(err: BaseException, msg: str) -> BaseException:
   ...:     """
   ...:     Adds a note to an exception inplace and returns it.
   ...:     """
   ...:     if sys.version_info < (3, 11):
   ...:         import exceptiongroup
   ...:         err.__notes__ = getattr(err, "__notes__", []) + [msg]
   ...:     else:
   ...:         err.add_note(msg)
   ...:     return err
   ...: 
   ...: 
   ...: with pytest.raises(Exception) as exc_info:
   ...:     e = ValueError("This is a test")
   ...:     add_note(e, "this is a note")
   ...:     raise e
   ...: 
   ...: traceback.format_exception_only(exc_info.type, exc_info.value)
Out[1]: ['ValueError: This is a test\n']
```